src/Avalonia.Controls/Expander.cs

```diff
@@ -37,60 +41,231 @@ public enum ExpandDirection
         {
-            UpdatePseudoClasses(ExpandDirection);
+            UpdatePseudoClasses();
         }
 
+        /// <summary>
+        /// Gets or sets the transition used when expanding or collapsing the content.
+        /// </summary>
         public IPageTransition? ContentTransition
         {
             get => GetValue(ContentTransitionProperty);
             set => SetValue(ContentTransitionProperty, value);
         }
 
+        /// <summary>
+        /// Gets or sets the direction in which the <see cref="Expander"/> opens.
+        /// </summary>
         public ExpandDirection ExpandDirection
         {
             get => GetValue(ExpandDirectionProperty);
             set => SetValue(ExpandDirectionProperty, value);
         }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether the <see cref="Expander"/>
+        /// content area is open and visible.
+        /// </summary>
         public bool IsExpanded
         {
-            get { return _isExpanded; }
-            set 
-            { 
-                SetAndRaise(IsExpandedProperty, ref _isExpanded, value);
-                PseudoClasses.Set(":expanded", value);
+            get => _isExpanded;
+            set
+            {
+                // It is important here that IsExpanded is a direct property so events can be invoked
+                // BEFORE the property system gets notified of updated values. This is because events
+                // may be canceled by external code.
+                if (_isExpanded != value)
+                {
+                    RoutedEventArgs eventArgs;
+
+                    if (value)
+                    {
+                        eventArgs = new RoutedEventArgs(ExpandingEvent, this);
+                        OnExpanding(eventArgs);
+                    }
+                    else
+                    {
+                        eventArgs = new RoutedEventArgs(CollapsingEvent, this);
+                        OnCollapsing(eventArgs);
+                    }
+
+                    if (eventArgs.Handled)
+                    {
+                        // If the event was externally handled (canceled) we must still notify the value has changed.
+                        // This property changed notification will update any external code observing this property that itself may have set the new value.
+                        // We are essentially reverted any external state change along with ignoring the IsExpanded property set.
+                        // Remember IsExpanded is usually controlled by a ToggleButton in the control theme.
+                        _ignorePropertyChanged = true;
+
+                        RaisePropertyChanged(
+                            IsExpandedProperty,
+                            oldValue: value,
+                            newValue: _isExpanded,
+                            BindingPriority.LocalValue,
+                            isEffectiveValue: true);
```

Comment:
This is kind of code I generally would prefer to avoid, as this API changes the property value without actually setting the value in conventional way.
Although, it also seems to the most convenient way to do these events for the users who would use this control.

Reply:
I tend to agree with you. I made sure to document reasoning well in case there are different suggestions. However, while it looks bad in general, I can't think of any downsides. It should be perfectly harmless. It's just one of those bad special cases.

That said, I did think of a way around this. It requires removing the toggle button entirely from the control template. State would then be managed entirely within the Expander itself -- which is probably a fundamentally better design to begin with. I think WPF was also being lazy and didn't want to duplicate a lot of code, as well.

Im not sure how much code duplication would be required to remove the toggle button from the default template. Pointer press and click would have to be entirely managed by the Expander. Keyboard navigation as well. It would probably be similar to what we did for SplitButton.

Reply:
I'll leave this open and we can always come back and change things in the future. It shouldn't be a bad breaking change except for those writing their own control themes (rare).

Reply:
It probably will be less customizable without the toggle button